rsmtool/rsmexplain.py

```diff
@@ -0,0 +1,525 @@
+        feature_names=explanation.feature_names,
+    )
+
+    # generate the HTML report
+    generate_report(explanation, output_dir, ids, configuration, logger)
+
+
+def generate_report(explanation, output_dir, ids, configuration, logger=None):
+    """
+    Generate an rsmexplain report.
+
+    This function also saves a series of files to disk, including
+    pickled versions of the explanation object and the ID dictionary.
+    All SHAP values are also saved as CSV files.
+
+    Parameters
+    ----------
+    explanation: shap.Explanation
+        SHAP explanation object containing shap values, data points, feature
+        names and base values.
+    output_dir : str
+        Path to the experiment output directory.
+    ids: dict
+        Dictionary mapping new row indices to original FeatureSet indices.
+    configuration: rsmtool.configuration_parser.Configuration
+        The Configuration object for rsmexplain.
+    logger : logging object, optional
+        A logging object. If ``None`` is passed, get logger from ``__name__``.
+        Defaults to ``None``.
+
+    """
+    logger = logger if logger else logging.getLogger(__name__)
+
+    # get the various output sub-directories which should already exist
+    csvdir = abspath(join(output_dir, "output"))
+    reportdir = abspath(join(output_dir, "report"))
+
+    # get the experiment ID
+    experiment_id = configuration["experiment_id"]
+
+    # first write the explanation object to disk, in case need it later
+    explanation_path = join(csvdir, f"{experiment_id}_explanation.pkl")
+    with open(explanation_path, "wb") as pickle_out:
+        pickle.dump(explanation, pickle_out)
+    configuration["explanation"] = explanation_path
+
+    id_path = join(csvdir, f"{experiment_id}_ids.pkl")
+    with open(id_path, "wb") as pickle_out:
+        pickle.dump(ids, pickle_out)
+    configuration["ids"] = id_path
+
+    # create various versions of the shap values to write to disk
+    csv_path = join(csvdir, f"{experiment_id}_shap_values.csv")
+    shap_frame = pd.DataFrame(
+        explanation.values, columns=explanation.feature_names, index=ids.values()
+    )
+    shap_frame.to_csv(csv_path)
+
+    # compute the various absolute value variants of the SHAP values and
+    # write out that dataframe to disk.
+    csv_path_abs = join(csvdir, f"{experiment_id}_absolute_shap_values.csv")
+    df_abs = pd.DataFrame(
+        [shap_frame.abs().mean(), shap_frame.abs().max(), shap_frame.abs().min()],
+        index=["abs. mean shap", "abs. max shap", "abs. min shap"],
+    ).transpose()
+    df_abs.to_csv(csv_path_abs, index_label="Feature")
+
+    # Initialize a reporter instance and add the sections:
+    reporter = Reporter(logger=logger)
+    general_report_sections = configuration["general_sections"]
+    special_report_sections = configuration["special_sections"]
+
+    # get any custom sections and locate them to make sure
+    # that they exist, otherwise raise an exception
```

Comment:
no exception is raised - will in come from the Reporter call?

Reply:
yep, that's right.